lib/Sema/CSDiag.cpp

```diff
@@ -4328,19 +4328,18 @@ bool FailureDiagnosis::visitExpr(Expr *E) {
   // independently invalid.
   bool errorInSubExpr = false;
   
-  E->forEachImmediateChildExpr([&](Expr *Child) -> Expr* {
+  E->forEachImmediateChildExpr([&](Expr *Child) {
     // If we already found an error, stop checking.
-    if (errorInSubExpr) return Child;
+    if (errorInSubExpr) return;
 
     // Otherwise just type check the subexpression independently.  If that
     // succeeds, then we stitch the result back into our expression.
     if (typeCheckChildIndependently(Child, TCC_AllowLValue))
```

Comment:
`typeCheckChildIndependently` could mutate AST which affects CSDiag diagnostics because some of the diagnostics there expect "type-checked" version of AST with types.

Reply:
I'm not changing that code, though. The return value is unused either way. (It'd normally be the next expression to visit, but because this method only visits immediate children it doesn't actually get used.)

Reply:
I haven't double-checked but I'm going to rely on what you are saying.  Would `ASTWalker` mutate AST with a result of `walkToExprPre` if it has to stop?

Reply:
ASTWalker never mutates the AST, and there's no more or less stopping than there was before.

Reply:
Alright, sounds good!

Reply:
This particular case doesn't change because `Child` is passed by value rather than by reference, so any mutations will happen below `Child`.